File: apps/reading/src/api/floatingResearchViewModeCompose.ts
/**
 * Floating research view-mode compose (pure).
 *
 * Operator vision: from a floating deep-research instance, choose float,
 * fullscreen, draft-merge intent, or full-merge intent — without live
 * dispatch or executed merges.
 *
 * Composes existing floatingDeepResearch helpers into one advisory snapshot.
 * live_dispatched and merge_executed are always false.
 */

import {
  proposeDraftMerge,
  proposeFullMerge,
  setFloatingViewMode,
  type FloatingDeepResearchInstance,
  type FloatingResearchViewMode,
  type MergeIntent,
} from "./floatingDeepResearch";

export type FloatingViewModeAction =
  | "float"
  | "fullscreen"
  | "propose_draft_merge"
  | "propose_full_merge";

export interface FloatingResearchViewModeComposeInput {
  instance: FloatingDeepResearchInstance;
  action: FloatingViewModeAction;
  /**
   * Required true for propose_full_merge (fail closed).
   * Ignored for float/fullscreen/draft (draft is preview intent).
   */
  operator_ack?: boolean;
}

export interface FloatingViewModeCapabilities {
  can_float: boolean;
  can_fullscreen: boolean;
  can_draft_merge: boolean;
  can_full_merge: boolean;
  current_view_mode: FloatingResearchViewMode;
  status: FloatingDeepResearchInstance["status"];
  notes: string[];
}

export interface FloatingResearchViewModeCompose {
  instance: FloatingDeepResearchInstance;
  action: FloatingViewModeAction;
  view_mode: FloatingResearchViewMode;
  merge_intent: MergeIntent | null;
  capabilities: FloatingViewModeCapabilities;
  /** True when action produced a valid next state / intent. */
  action_applied: boolean;
  /** Always false — pure layer never dispatches providers. */
  live_dispatched: false;
  /** Always false — pure layer never mutates parent assets. */
  merge_executed: false;
  notes: string[];
  authority: "floating_research_view_mode_compose_advisory";
}

function requireInstance(
  instance: unknown,
): FloatingDeepResearchInstance {
  if (!instance || typeof instance !== "object") {
    throw new Error("instance must be an object");
  }
  const inst = instance as FloatingDeepResearchInstance;
  if (inst.live_dispatched !== false) {
    throw new Error("live_dispatched must be false (pure layer)");
  }
  if (inst.merge_executed !== false) {
    throw new Error("merge_executed must be false (pure layer)");
  }
  if (typeof inst.instance_id !== "string" || !inst.instance_id.trim()) {
    throw new Error("instance.instance_id must be a non-empty string");
  }
  if (
    typeof inst.parent_asset_id !== "string" ||
    !inst.parent_asset_id.trim()
  ) {
    throw new Error("instance.parent_asset_id must be a non-empty string");
  }
  return inst;
}

/**
 * Assess what view-mode actions are valid for this pure instance.
 * Never dispatches or merges.
 */
export function assessFloatingViewModeCapabilities(
  instance: FloatingDeepResearchInstance,
): FloatingViewModeCapabilities {
  const inst = requireInstance(instance);
  const notes: string[] = [
    "capabilities are pure advisory — no live dispatch",
    "live_dispatched=false",
    "merge_executed=false",
  ];

  const closed = inst.status === "closed";
  // Float / fullscreen always available when not closed and honesty flags hold.
  const can_float = !closed;
  const can_fullscreen = !closed;
  // Draft merge intent allowed for proposed|open|completed (mirrors proposeDraftMerge).
  const can_draft_merge =
    !closed &&
    (inst.status === "proposed" ||
      inst.status === "open" ||
      inst.status === "completed");
  // Full merge only when completed (operator_ack still required at compose time).
  const can_full_merge = !closed && inst.status === "completed";

  if (closed) {
    notes.push("status=closed — no view-mode actions");
  } else {
    notes.push(
      `can_float=${can_float} can_fullscreen=${can_fullscreen} ` +
        `can_draft_merge=${can_draft_merge} can_full_merge=${can_full_merge}`,
    );
  }

  return {
    can_float,
    can_fullscreen,
    can_draft_merge,
    can_full_merge,
    current_view_mode: inst.view_mode,
    status: inst.status,
    notes,
  };
}

/**
 * Apply a pure view-mode action (float / fullscreen / draft|full merge intent).
 * Never live-dispatches; never executes merges.
 */
export function composeFloatingResearchViewMode(
  input: FloatingResearchViewModeComposeInput,
): FloatingResearchViewModeCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  const inst0 = requireInstance(input.instance);
  const action = input.action;
  if (
    action !== "float" &&
    action !== "fullscreen" &&
    action !== "propose_draft_merge" &&
    action !== "propose_full_merge"
  ) {
    throw new Error(
      "action must be float|fullscreen|propose_draft_merge|propose_full_merge",
    );
  }

  const capabilities = assessFloatingViewModeCapabilities(inst0);
  const notes: string[] = [
    "live_dispatched=false — no provider dispatch from view-mode compose",
    "merge_executed=false — parent asset never mutated in pure layer",
    ...capabilities.notes,
  ];

  let instance = inst0;
  let merge_intent: MergeIntent | null = null;
  let action_applied = false;

  if (action === "float") {
    if (!capabilities.can_float) {
      throw new Error("cannot float: instance closed or invalid");
    }
    instance = setFloatingViewMode(inst0, "floating");
    action_applied = true;
    notes.push("action=float → view_mode=floating");
  } else if (action === "fullscreen") {
    if (!capabilities.can_fullscreen) {
      throw new Error("cannot fullscreen: instance closed or invalid");
    }
    instance = setFloatingViewMode(inst0, "fullscreen");
    action_applied = true;
    notes.push("action=fullscreen → view_mode=fullscreen");
  } else if (action === "propose_draft_merge") {
    if (!capabilities.can_draft_merge) {
      throw new Error(
        "cannot propose draft merge: requires proposed|open|completed",
      );
    }
    merge_intent = proposeDraftMerge(inst0);
    // View mode stays non-merged; merge modes only via intent helpers.
    action_applied = true;
    notes.push(
      "action=propose_draft_merge · merge_intent kind=draft_merge · merge_executed=false",
    );
  } else {
    // propose_full_merge
    if (!capabilities.can_full_merge) {
      throw new Error("cannot propose full merge: requires completed status");
    }
    if (typeof input.operator_ack !== "boolean") {
      throw new Error(
        "operator_ack must be an explicit boolean for propose_full_merge",
      );
    }
    merge_intent = proposeFullMerge(inst0, {
      operator_ack: input.operator_ack,
    });
    action_applied = true;
    notes.push(
      "action=propose_full_merge · merge_intent kind=full_merge · merge_executed=false",
    );
  }

  // Re-assert honesty flags after helper chain.
  if (instance.live_dispatched !== false) {
    throw new Error("invariant: live_dispatched must remain false");
  }
  if (instance.merge_executed !== false) {
    throw new Error("invariant: merge_executed must remain false");
  }
  if (merge_intent != null && merge_intent.merge_executed !== false) {
    throw new Error("invariant: merge_intent.merge_executed must be false");
  }

  notes.push("live_dispatched=false");
  notes.push("merge_executed=false");

  return {
    instance,
    action,
    view_mode: instance.view_mode,
    merge_intent,
    capabilities,
    action_applied,
    live_dispatched: false,
    merge_executed: false,
    notes,
    authority: "floating_research_view_mode_compose_advisory",
  };
}

export function formatFloatingViewModeComposeSummary(
  c: FloatingResearchViewModeCompose,
): string {
  const intent = c.merge_intent ? c.merge_intent.kind : "none";
  return (
    `action=${c.action} · view_mode=${c.view_mode} · intent=${intent} · ` +
    `applied=${c.action_applied} · live_dispatched=false · merge_executed=false`
  );
}

File: substrate/floating_research_view_mode_compose.py
"""Floating research view-mode compose (pure).

Operator vision: from a floating deep-research instance, choose float,
fullscreen, draft-merge intent, or full-merge intent — without live
dispatch or executed merges.

Composes substrate.floating_deep_research helpers into one advisory snapshot.
live_dispatched and merge_executed are always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from substrate.floating_deep_research import (
    FloatingDeepResearchError,
    FloatingDeepResearchInstance,
    MergeIntent,
    mark_floating_completed,
    propose_draft_merge,
    propose_full_merge,
    set_floating_view_mode,
    spawn_floating_from_highlight,
)

ViewModeAction = Literal[
    "float",
    "fullscreen",
    "propose_draft_merge",
    "propose_full_merge",
]

VALID_ACTIONS: frozenset[str] = frozenset(
    ("float", "fullscreen", "propose_draft_merge", "propose_full_merge")
)


class FloatingResearchViewModeComposeError(ValueError):
    """Fail-closed validation for view-mode compose."""


@dataclass(frozen=True)
class FloatingViewModeCapabilities:
    can_float: bool
    can_fullscreen: bool
    can_draft_merge: bool
    can_full_merge: bool
    current_view_mode: str
    status: str
    notes: tuple[str, ...]

    def to_dict(self) -> dict[str, Any]:
        return {
            "can_float": self.can_float,
            "can_fullscreen": self.can_fullscreen,
            "can_draft_merge": self.can_draft_merge,
            "can_full_merge": self.can_full_merge,
            "current_view_mode": self.current_view_mode,
            "status": self.status,
            "notes": list(self.notes),
        }


@dataclass(frozen=True)
class FloatingResearchViewModeCompose:
    instance: FloatingDeepResearchInstance
    action: str
    view_mode: str
    merge_intent: MergeIntent | None
    capabilities: FloatingViewModeCapabilities
    action_applied: bool
    live_dispatched: bool
    merge_executed: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "instance": self.instance.to_dict(),
            "action": self.action,
            "view_mode": self.view_mode,
            "merge_intent": (
                self.merge_intent.to_dict() if self.merge_intent is not None else None
            ),
            "capabilities": self.capabilities.to_dict(),
            "action_applied": self.action_applied,
            "live_dispatched": False,
            "merge_executed": False,
            "notes": list(self.notes),
            "authority": "floating_research_view_mode_compose_advisory",
        }


def _require_instance(instance: object) -> FloatingDeepResearchInstance:
    if not isinstance(instance, FloatingDeepResearchInstance):
        raise FloatingResearchViewModeComposeError(
            "instance must be FloatingDeepResearchInstance"
        )
    if instance.live_dispatched is not False:
        raise FloatingResearchViewModeComposeError(
            "live_dispatched must be false (pure layer)"
        )
    if instance.merge_executed is not False:
        raise FloatingResearchViewModeComposeError(
            "merge_executed must be false (pure layer)"
        )
    return instance


def assess_floating_view_mode_capabilities(
    instance: FloatingDeepResearchInstance,
) -> FloatingViewModeCapabilities:
    """Assess valid pure view-mode actions. Never dispatches or merges."""
    inst = _require_instance(instance)
    notes: list[str] = [
        "capabilities are pure advisory — no live dispatch",
        "live_dispatched=false",
        "merge_executed=false",
    ]
    closed = inst.status == "closed"
    can_float = not closed
    can_fullscreen = not closed
    can_draft_merge = (not closed) and inst.status in (
        "proposed",
        "open",
        "completed",
    )
    can_full_merge = (not closed) and inst.status == "completed"
    if closed:
        notes.append("status=closed — no view-mode actions")
    else:
        notes.append(
            f"can_float={can_float} can_fullscreen={can_fullscreen} "
            f"can_draft_merge={can_draft_merge} can_full_merge={can_full_merge}"
        )
    return FloatingViewModeCapabilities(
        can_float=can_float,
        can_fullscreen=can_fullscreen,
        can_draft_merge=can_draft_merge,
        can_full_merge=can_full_merge,
        current_view_mode=inst.view_mode,
        status=inst.status,
        notes=tuple(notes),
    )


def compose_floating_research_view_mode(
    *,
    instance: FloatingDeepResearchInstance,
    action: object,
    operator_ack: object | None = None,
) -> FloatingResearchViewModeCompose:
    """Apply pure view-mode action. Never live-dispatches; never merges."""
    inst0 = _require_instance(instance)
    if not isinstance(action, str) or action not in VALID_ACTIONS:
        raise FloatingResearchViewModeComposeError(
            "action must be float|fullscreen|propose_draft_merge|propose_full_merge"
        )
    capabilities = assess_floating_view_mode_capabilities(inst0)
    notes: list[str] = [
        "live_dispatched=false — no provider dispatch from view-mode compose",
        "merge_executed=false — parent asset never mutated in pure layer",
        *capabilities.notes,
    ]

    inst = inst0
    merge_intent: MergeIntent | None = None
    action_applied = False

    try:
        if action == "float":
            if not capabilities.can_float:
                raise FloatingResearchViewModeComposeError(
                    "cannot float: instance closed or invalid"
                )
            inst = set_floating_view_mode(inst0, "floating")
            action_applied = True
            notes.append("action=float → view_mode=floating")
        elif action == "fullscreen":
            if not capabilities.can_fullscreen:
                raise FloatingResearchViewModeComposeError(
                    "cannot fullscreen: instance closed or invalid"
                )
            inst = set_floating_view_mode(inst0, "fullscreen")
            action_applied = True
            notes.append("action=fullscreen → view_mode=fullscreen")
        elif action == "propose_draft_merge":
            if not capabilities.can_draft_merge:
                raise FloatingResearchViewModeComposeError(
                    "cannot propose draft merge: requires proposed|open|completed"
                )
            merge_intent = propose_draft_merge(inst0)
            action_applied = True
            notes.append(
                "action=propose_draft_merge · merge_intent kind=draft_merge · "
                "merge_executed=false"
            )
        else:
            # propose_full_merge
            if not capabilities.can_full_merge:
                raise FloatingResearchViewModeComposeError(
                    "cannot propose full merge: requires completed status"
                )
            if not isinstance(operator_ack, bool):
                raise FloatingResearchViewModeComposeError(
                    "operator_ack must be an explicit boolean for propose_full_merge"
                )
            merge_intent = propose_full_merge(inst0, operator_ack=operator_ack)
            action_applied = True
            notes.append(
                "action=propose_full_merge · merge_intent kind=full_merge · "
                "merge_executed=false"
            )
    except FloatingDeepResearchError as e:
        raise FloatingResearchViewModeComposeError(str(e)) from e

    if inst.live_dispatched is not False or inst.merge_executed is not False:
        raise FloatingResearchViewModeComposeError(
            "invariant: honesty flags must remain false on instance"
        )
    if merge_intent is not None and merge_intent.merge_executed is not False:
        raise FloatingResearchViewModeComposeError(
            "invariant: merge_intent.merge_executed must be false"
        )

    notes.append("live_dispatched=false")
    notes.append("merge_executed=false")

    return FloatingResearchViewModeCompose(
        instance=inst,
        action=action,
        view_mode=inst.view_mode,
        merge_intent=merge_intent,
        capabilities=capabilities,
        action_applied=action_applied,
        live_dispatched=False,
        merge_executed=False,
        notes=tuple(notes),
        authority="floating_research_view_mode_compose_advisory",
    )


def format_floating_view_mode_compose_summary(
    c: FloatingResearchViewModeCompose,
) -> str:
    intent = c.merge_intent.kind if c.merge_intent is not None else "none"
    return (
        f"action={c.action} · view_mode={c.view_mode} · intent={intent} · "
        f"applied={c.action_applied} · live_dispatched=false · merge_executed=false"
    )


# Re-export spawn helpers for routes that accept raw highlight fields.
__all__ = [
    "FloatingResearchViewModeCompose",
    "FloatingResearchViewModeComposeError",
    "FloatingViewModeCapabilities",
    "VALID_ACTIONS",
    "assess_floating_view_mode_capabilities",
    "compose_floating_research_view_mode",
    "format_floating_view_mode_compose_summary",
    "mark_floating_completed",
    "spawn_floating_from_highlight",
]
